logging/writer.go

```diff
@@ -1,29 +1,90 @@
 package logging
 
-import "io"
+import (
+	"bytes"
+	"io"
+	"strings"
+	"sync"
+)
 
 // writer is an implementation of io.Writer that writes to a logger.
 type writer struct {
 	l Logger
 	f func(Logger, string)
+
+	m   sync.Mutex
+	buf strings.Builder
 }
 
-// NewWriter returns an io.Writer that writes content to the given logger.
+// NewWriteCloser returns an io.WriteCloser that writes content to the given
+// logger.
+//
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
 //
-// Each call to Write() produces a separate log message.
-func NewWriter(l Logger) io.Writer {
-	return &writer{l, LogString}
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
+//
+// When Close() is called the writer checks if there is any remaining buffered
+// string content available and, if there is, the writer flushes the remaining
+// content as a message to the Logger.
+func NewWriteCloser(l Logger) io.WriteCloser {
```

Comment:
```suggestion
func NewWriter(l Logger) io.WriteCloser {
```

I'd prefer not to change these function names and cause a BC break, as they'll behave quite similar even with the changes we've put forward.

Reply:
Yeah I thought about it too. Fair point again

Reply:
Yeah man, I feel that docs are a bit bulky. I definitely welcome any suggestion to shrink them.